### ISSUE TITLE
Depends: libreadline6 (>= 6.3) but it is not installable

### ISSUE THREAD

**Issue by borrascador** [mellowcandle/bitwise#11]
Trying to install and I'm getting the following error. I can't seem to find libreadline6 using `apt-get install libreadline6` either.

Running Ubuntu 18.04.2 on WSL.

```
$ sudo apt-get install bitwise

Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 bitwise : Depends: libreadline6 (>= 6.3) but it is not installable
E: Unable to correct problems, you have held broken packages.
```

**Comment by mellowcandle** [comment]
Try this:
http://zoomadmin.com/HowToInstall/UbuntuPackage/libreadline7

**Comment by borrascador** [comment]
After running `sudo apt-get -y install libreadline7` I get `libreadline7 is already the newest version (7.0-3).`

**Comment by borrascador** [comment]
It appears that I am having the same problem as this user: https://unix.stackexchange.com/questions/467020/libreadline-dependency-related-error-while-installing-an-app-seedsync

**Comment by mellowcandle** [comment]
I suggest you try to use snap for installing bitwise.

**Comment by stevemk14ebr** [comment]
First off, i suggest debian not ubuntu for WSL. But try to install: 

```
sudo apt-get install libreadline-dev 
```

**Comment by Chemsmith** [comment]
> I suggest you try to use snap for installing bitwise.
snap currently does not work under WSL :(

> First off, i suggest debian not ubuntu for WSL. But try to install:
> 
> ```
> sudo apt-get install libreadline-dev 
> ```
That does not assist in installing bitwise from the ppa.

Ubuntu 19.04 has libreadline8 installed.

**Comment by stevemk14ebr** [comment]
oops sorry, that for source. Build it from source it's easy https://github.com/mellowcandle/bitwise/pull/19/commits/a8219c382c3b4ae4dcc3f618b21f49223f800c46

**Comment by jetzerb** [comment]
Looks like the dependencies in the `deb` include a version of `libreadline` that isn't available on recent versions of Ubuntu.  I'm running 19.04 and got the same error message when installing.

The dependencies of the package are listed as
`Depends: libc6 (>= 2.7), libncurses5 (>= 5.5-5~), libreadline6 (>= 6.3), libtinfo5, libreadline5`

@mellowcandle : is that correct?  `bitwise` requires both `libreadline6` _and_ `libreadline5`?

All of the dependencies can be met on both Ubuntu 18.04 (Bionic) and 19.04 (Disco), except for `libreadline6`.  18.04 has only `libreadline`**7**, while 19.04 has only `libreadline`**8**.

`bitwise` can still be installed though, if you force it:

```
# become root
sudo -i

# ensure all the other dependencies are installed
apt-get install libc6 libncurses5 librtinfo5 libreadline5

# manually download the bitwise package and force the installation
cd /tmp
apt-get download bitwise
dpkg --force-depends -i bitwise_0.30-2ubuntu3_amd64.deb
# ...here, dpkg tells you about the missing dependency...

# Create a libreadline.so.6 symlink from your installed version
cd /lib/x86_64-linux-gnu # see output of dpkg -L libreadline5 for actual directory
ln -s libreadline.so.[78] libreadline.so.6
```

You can now run bitwise, and AFAICT, it's fully functional.

**Comment by mellowcandle** [comment]
creating Debian packages is a pain. what initially do, is to compile for trusty and then copy the binaries to all other versions. this is the recommended version by launchpad.

**Comment by mellowcandle** [comment]
I think it will be better to start build on a more recent version, who uses trusty nowadays, I think it's 12.04

**Comment by mellowcandle** [comment]
I will try with newest version v0.31  that I released today. hopefully it will work better.

**Comment by mellowcandle** [comment]
> The dependencies of the package are listed as
> `Depends: libc6 (>= 2.7), libncurses5 (>= 5.5-5~), libreadline6 (>= 6.3), libtinfo5, libreadline5`
> 
> @mellowcandle : is that correct? `bitwise` requires both `libreadline6` _and_ `libreadline5`?

This are the lines that define the dependencies:
```
Build-Depends: debhelper (>= 9), autotools-dev, libncurses5-dev, libreadline-dev, libcunit1-dev
Depends: ${shlibs:Depends}, ${misc:Depends}, libncurses5, libreadline5
```

I don't define libreadline6 nowhere.
It's weird.

**Comment by jetzerb** [comment]
I've got roughly zero experience building `deb` packages, but I'd guess that your `libreadline-dev` build dependency turns into a specific `libreadline`_n_ dependency for the output package.

**Ubuntu 18.04 / Bionic:**
```
$ apt-cache madison libreadline-dev
libreadline-dev |      7.0-3 | http://archive.ubuntu.com/ubuntu bionic/main amd64 Packages
```

**Ubuntu 19.04 / Disco:**
```
apt-cache madison libreadline-dev
libreadline-dev |      8.0-1 | http://archive.ubuntu.com/ubuntu disco/main amd64 Packages
```

My guess would be that Trusty (where you said you compiled it) has libreadline-dev 6.x, so at runtime it requires `libreadline.so.6`.

I wonder if you really need libreadline5 at all...
```
$ strings $(which bitwise) | grep libreadline
libreadline.so.6
```
Nothing in there about v5...

**Comment by mellowcandle** [comment]
are you all on WSL? because I managed to install it on at least two 18.04 desktops.

**Comment by jetzerb** [comment]
I'm running inside of [Docker containers](https://hub.docker.com/_/ubuntu).

**Comment by mellowcandle** [comment]
on Monday I'll try to catch some launchpad devs on irc, I'll be next to ubuntu at work so I could debug it.

**Comment by borrascador** [comment]
@stevemk14ebr I'll try to install this from source on WSL next week and report back here.

**Comment by mellowcandle** [comment]
@jetzerb  @stevemk14ebr @borrascador 
I think it should be resolved now.
I build v0.32 for each release and didn't copy forward. please `apt-get update` and `apt-get install bitwise`.

Please report back if problem is resolved.

Thanks.

**Comment by jetzerb** [comment]
**I verified that it installs properly on both Ubuntu 18.04 and 19.04.
Thank you for your efforts!**

**Comment by borrascador** [comment]
Thanks for the fix -- verified that it works on Ubuntu 18.04 running on WSL in Windows 10.
